### PR TITLE
Add new campaign rules and add functionality to campaigns

### PR DESCRIPTION
**Pull request by Pikkupomo**
Add quantities to existing rules
Campaigns: Add product discount amount effect
Enable multiple effects for campaigns
Show quantity in basket for all lines

Refs SHUUP-2733, SHUUP-2734, SHUUP-2735

### CODE REVIEW

File: shoop/campaigns/modules.py [hunk 1]
```diff
@@ -111,21 +111,21 @@ def get_discount_line(campaign, amount, price_so_far):
...
-            # if campaign has coupon, match it to order_source.codes
-            if campaign.coupon:
-                # campaign was found because discount code matched. This line is always added
-                lines.append(get_discount_line(campaign, discount_amount, price_so_far))
-            elif best_discount is None or discount_amount > best_discount:
-                best_discount = discount_amount
-                best_discount_campaign = campaign
+            for effect in campaign.effects.all():
+                if not effect:
+                    continue
```
Comment (shawnadelic) [inline shoop/campaigns/modules.py]: I don't think this is necessary in the inner loop.

File: shoop/campaigns/modules.py [hunk 2]
```diff
@@ -134,12 +134,12 @@ def get_discount_line(campaign, amount, price_so_far):
     def _handle_line_effects(self, matching_campaigns, order_source, original_lines):
         lines = []
         for campaign in matching_campaigns:
-            effect = campaign.effects.first()
-            if not effect:
-                continue
+            for effect in campaign.effects.all():
+                if not effect:
+                    continue
```
Comment (shawnadelic) [inline shoop/campaigns/modules.py]: Again, not sure if this is necessary

File: shoop/campaigns/modules.py [hunk 3]
```diff
@@ -111,21 +111,21 @@ def get_discount_line(campaign, amount, price_so_far):
...
-                lines.append(get_discount_line(campaign, discount_amount, price_so_far))
-            elif best_discount is None or discount_amount > best_discount:
-                best_discount = discount_amount
-                best_discount_campaign = campaign
+            for effect in campaign.effects.all():
+                if not effect:
+                    continue
+
+                if hasattr(effect, "get_discount_lines"):
+                    continue
```
Comment (suutari-ai) [inline shoop/campaigns/modules.py]: Why are you mixing two kinds of effects in the same list? If the BasketLineEffects and BasketDiscountEffects were using two different related names, then there would be no need for this hasattr hack.

File: shoop/campaigns/modules.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/campaigns/modules.py]: Yep, not needed.

File: shoop/campaigns/modules.py (hunk 3 above)
Comment (Pikkupomo) [inline shoop/campaigns/modules.py]: That's true, maybe I could rewrite this to match that schema.

**Pull request closed without merging**